{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "RFPlasmid systemwide install with pip",
      "provenance": [],
      "authorship_tag": "ABX9TyN6dIoEsgPYHgWOgQLhfyPY",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/aldertzomer/RFPlasmid/blob/master/RFPlasmid_systemwide_install_with_pip.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "IZalEpa0u5YJ"
      },
      "source": [
        "# Welcome!\n",
        "\n",
        "This Google Colab script shows how to install rfplasmid systemwide using pip and allows using it as well by mounting your Google drive and running it on a folder there. Please send adjustments and improvements to a.l.zomer@uu.nl\n"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "aq2HkZzWe0hV",
        "outputId": "3ce752f2-96b8-48cb-b6bc-9c72c254daa4"
      },
      "source": [
        "#DIAMOND is required for RFPlasmid\n",
        "%%shell\n",
        "wget http://github.com/bbuchfink/diamond/releases/download/v0.9.24/diamond-linux64.tar.gz\n",
        "tar xzf diamond-linux64.tar.gz\n",
        "cp diamond /usr/local/bin/diamond\n",
        "diamond --version"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "--2021-06-07 13:58:56--  http://github.com/bbuchfink/diamond/releases/download/v0.9.24/diamond-linux64.tar.gz\n",
            "Resolving github.com (github.com)... 140.82.113.3\n",
            "Connecting to github.com (github.com)|140.82.113.3|:80... connected.\n",
            "HTTP request sent, awaiting response... 301 Moved Permanently\n",
            "Location: https://github.com/bbuchfink/diamond/releases/download/v0.9.24/diamond-linux64.tar.gz [following]\n",
            "--2021-06-07 13:58:56--  https://github.com/bbuchfink/diamond/releases/download/v0.9.24/diamond-linux64.tar.gz\n",
            "Connecting to github.com (github.com)|140.82.113.3|:443... connected.\n",
            "HTTP request sent, awaiting response... 302 Found\n",
            "Location: https://github-releases.githubusercontent.com/31987083/313cc780-09dd-11e9-902e-599c1618e37d?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20210607%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20210607T135857Z&X-Amz-Expires=300&X-Amz-Signature=a680a89b6b192b6e6116315ae6c4237782a71ae8067a4188b4d28fd8e6122391&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=31987083&response-content-disposition=attachment%3B%20filename%3Ddiamond-linux64.tar.gz&response-content-type=application%2Foctet-stream [following]\n",
            "--2021-06-07 13:58:57--  https://github-releases.githubusercontent.com/31987083/313cc780-09dd-11e9-902e-599c1618e37d?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20210607%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20210607T135857Z&X-Amz-Expires=300&X-Amz-Signature=a680a89b6b192b6e6116315ae6c4237782a71ae8067a4188b4d28fd8e6122391&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=31987083&response-content-disposition=attachment%3B%20filename%3Ddiamond-linux64.tar.gz&response-content-type=application%2Foctet-stream\n",
            "Resolving github-releases.githubusercontent.com (github-releases.githubusercontent.com)... 185.199.108.154, 185.199.109.154, 185.199.110.154, ...\n",
            "Connecting to github-releases.githubusercontent.com (github-releases.githubusercontent.com)|185.199.108.154|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 2418573 (2.3M) [application/octet-stream]\n",
            "Saving to: ‘diamond-linux64.tar.gz’\n",
            "\n",
            "diamond-linux64.tar 100%[===================>]   2.31M  --.-KB/s    in 0.04s   \n",
            "\n",
            "2021-06-07 13:58:57 (62.4 MB/s) - ‘diamond-linux64.tar.gz’ saved [2418573/2418573]\n",
            "\n",
            "diamond version 0.9.24\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              ""
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 1
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "gausp5B2qXWZ",
        "outputId": "11d4881a-73af-47da-bb12-e8103b818e96"
      },
      "source": [
        "#CheckM requires HMMER\n",
        "%%shell\n",
        "apt install hmmer\n",
        "hmmsearch -h"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Reading package lists... Done\n",
            "Building dependency tree       \n",
            "Reading state information... Done\n",
            "The following additional packages will be installed:\n",
            "  libdivsufsort3\n",
            "Suggested packages:\n",
            "  hmmer-doc\n",
            "The following NEW packages will be installed:\n",
            "  hmmer libdivsufsort3\n",
            "0 upgraded, 2 newly installed, 0 to remove and 39 not upgraded.\n",
            "Need to get 1,164 kB of archives.\n",
            "After this operation, 11.9 MB of additional disk space will be used.\n",
            "Get:1 http://archive.ubuntu.com/ubuntu bionic/universe amd64 libdivsufsort3 amd64 2.0.1-3 [44.4 kB]\n",
            "Get:2 http://archive.ubuntu.com/ubuntu bionic/universe amd64 hmmer amd64 3.1b2+dfsg-5ubuntu1 [1,119 kB]\n",
            "Fetched 1,164 kB in 1s (1,579 kB/s)\n",
            "Selecting previously unselected package libdivsufsort3:amd64.\n",
            "(Reading database ... 160772 files and directories currently installed.)\n",
            "Preparing to unpack .../libdivsufsort3_2.0.1-3_amd64.deb ...\n",
            "Unpacking libdivsufsort3:amd64 (2.0.1-3) ...\n",
            "Selecting previously unselected package hmmer.\n",
            "Preparing to unpack .../hmmer_3.1b2+dfsg-5ubuntu1_amd64.deb ...\n",
            "Unpacking hmmer (3.1b2+dfsg-5ubuntu1) ...\n",
            "Setting up libdivsufsort3:amd64 (2.0.1-3) ...\n",
            "Setting up hmmer (3.1b2+dfsg-5ubuntu1) ...\n",
            "Processing triggers for man-db (2.8.3-2ubuntu0.1) ...\n",
            "Processing triggers for libc-bin (2.27-3ubuntu1.2) ...\n",
            "/sbin/ldconfig.real: /usr/local/lib/python3.7/dist-packages/ideep4py/lib/libmkldnn.so.0 is not a symbolic link\n",
            "\n",
            "# hmmsearch :: search profile(s) against a sequence database\n",
            "# HMMER 3.1b2 (February 2015); http://hmmer.org/\n",
            "# Copyright (C) 2015 Howard Hughes Medical Institute.\n",
            "# Freely distributed under the GNU General Public License (GPLv3).\n",
            "# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -\n",
            "Usage: hmmsearch [options] <hmmfile> <seqdb>\n",
            "\n",
            "Basic options:\n",
            "  -h : show brief help on version and usage\n",
            "\n",
            "Options directing output:\n",
            "  -o <f>           : direct output to file <f>, not stdout\n",
            "  -A <f>           : save multiple alignment of all hits to file <f>\n",
            "  --tblout <f>     : save parseable table of per-sequence hits to file <f>\n",
            "  --domtblout <f>  : save parseable table of per-domain hits to file <f>\n",
            "  --pfamtblout <f> : save table of hits and domains to file, in Pfam format <f>\n",
            "  --acc            : prefer accessions over names in output\n",
            "  --noali          : don't output alignments, so output is smaller\n",
            "  --notextw        : unlimit ASCII text output line width\n",
            "  --textw <n>      : set max width of ASCII text output lines  [120]  (n>=120)\n",
            "\n",
            "Options controlling reporting thresholds:\n",
            "  -E <x>     : report sequences <= this E-value threshold in output  [10.0]  (x>0)\n",
            "  -T <x>     : report sequences >= this score threshold in output\n",
            "  --domE <x> : report domains <= this E-value threshold in output  [10.0]  (x>0)\n",
            "  --domT <x> : report domains >= this score cutoff in output\n",
            "\n",
            "Options controlling inclusion (significance) thresholds:\n",
            "  --incE <x>    : consider sequences <= this E-value threshold as significant\n",
            "  --incT <x>    : consider sequences >= this score threshold as significant\n",
            "  --incdomE <x> : consider domains <= this E-value threshold as significant\n",
            "  --incdomT <x> : consider domains >= this score threshold as significant\n",
            "\n",
            "Options controlling model-specific thresholding:\n",
            "  --cut_ga : use profile's GA gathering cutoffs to set all thresholding\n",
            "  --cut_nc : use profile's NC noise cutoffs to set all thresholding\n",
            "  --cut_tc : use profile's TC trusted cutoffs to set all thresholding\n",
            "\n",
            "Options controlling acceleration heuristics:\n",
            "  --max    : Turn all heuristic filters off (less speed, more power)\n",
            "  --F1 <x> : Stage 1 (MSV) threshold: promote hits w/ P <= F1  [0.02]\n",
            "  --F2 <x> : Stage 2 (Vit) threshold: promote hits w/ P <= F2  [1e-3]\n",
            "  --F3 <x> : Stage 3 (Fwd) threshold: promote hits w/ P <= F3  [1e-5]\n",
            "  --nobias : turn off composition bias filter\n",
            "\n",
            "Other expert options:\n",
            "  --nonull2     : turn off biased composition score corrections\n",
            "  -Z <x>        : set # of comparisons done, for E-value calculation\n",
            "  --domZ <x>    : set # of significant seqs, for domain E-value calculation\n",
            "  --seed <n>    : set RNG seed to <n> (if 0: one-time arbitrary seed)  [42]\n",
            "  --tformat <s> : assert target <seqfile> is in format <s>: no autodetection\n",
            "  --cpu <n>     : number of parallel CPU workers to use for multithreads\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              ""
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 2
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "ZGJj7ze1layv",
        "outputId": "9a60ad37-f0a5-4c75-cbd9-b9ad93ad5656"
      },
      "source": [
        "#CheckM requires Prodigal\n",
        "%%shell\n",
        "wget https://github.com/hyattpd/Prodigal/releases/download/v2.6.3/prodigal.linux\n",
        "cp prodigal.linux /usr/local/bin/prodigal\n",
        "chmod +x /usr/local/bin/prodigal\n",
        "prodigal"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "--2021-06-07 13:59:18--  https://github.com/hyattpd/Prodigal/releases/download/v2.6.3/prodigal.linux\n",
            "Resolving github.com (github.com)... 140.82.113.4\n",
            "Connecting to github.com (github.com)|140.82.113.4|:443... connected.\n",
            "HTTP request sent, awaiting response... 302 Found\n",
            "Location: https://github-releases.githubusercontent.com/11352636/bd77d970-d4ce-11e5-966e-1679b1b1d962?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20210607%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20210607T135915Z&X-Amz-Expires=300&X-Amz-Signature=ce5b4a76d4fddcaeaf088c5b5d413ed600f5f6cf1e2c1f55ab3b9763bbe95c9f&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=11352636&response-content-disposition=attachment%3B%20filename%3Dprodigal.linux&response-content-type=application%2Foctet-stream [following]\n",
            "--2021-06-07 13:59:18--  https://github-releases.githubusercontent.com/11352636/bd77d970-d4ce-11e5-966e-1679b1b1d962?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20210607%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20210607T135915Z&X-Amz-Expires=300&X-Amz-Signature=ce5b4a76d4fddcaeaf088c5b5d413ed600f5f6cf1e2c1f55ab3b9763bbe95c9f&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=11352636&response-content-disposition=attachment%3B%20filename%3Dprodigal.linux&response-content-type=application%2Foctet-stream\n",
            "Resolving github-releases.githubusercontent.com (github-releases.githubusercontent.com)... 185.199.110.154, 185.199.109.154, 185.199.108.154, ...\n",
            "Connecting to github-releases.githubusercontent.com (github-releases.githubusercontent.com)|185.199.110.154|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 13599280 (13M) [application/octet-stream]\n",
            "Saving to: ‘prodigal.linux.1’\n",
            "\n",
            "\rprodigal.linux.1      0%[                    ]       0  --.-KB/s               \rprodigal.linux.1    100%[===================>]  12.97M  --.-KB/s    in 0.09s   \n",
            "\n",
            "2021-06-07 13:59:19 (145 MB/s) - ‘prodigal.linux.1’ saved [13599280/13599280]\n",
            "\n",
            "-------------------------------------\n",
            "PRODIGAL v2.6.3 [February, 2016]         \n",
            "Univ of Tenn / Oak Ridge National Lab\n",
            "Doug Hyatt, Loren Hauser, et al.     \n",
            "-------------------------------------\n",
            "\n",
            "Usage:  prodigal [-a trans_file] [-c] [-d nuc_file] [-f output_type]\n",
            "                 [-g tr_table] [-h] [-i input_file] [-m] [-n] [-o output_file]\n",
            "                 [-p mode] [-q] [-s start_file] [-t training_file] [-v]\n",
            "\n",
            "         -a:  Write protein translations to the selected file.\n",
            "         -c:  Closed ends.  Do not allow genes to run off edges.\n",
            "         -d:  Write nucleotide sequences of genes to the selected file.\n",
            "         -f:  Select output format (gbk, gff, or sco).  Default is gbk.\n",
            "         -g:  Specify a translation table to use (default 11).\n",
            "         -h:  Print help menu and exit.\n",
            "         -i:  Specify FASTA/Genbank input file (default reads from stdin).\n",
            "         -m:  Treat runs of N as masked sequence; don't build genes across them.\n",
            "         -n:  Bypass Shine-Dalgarno trainer and force a full motif scan.\n",
            "         -o:  Specify output file (default writes to stdout).\n",
            "         -p:  Select procedure (single or meta).  Default is single.\n",
            "         -q:  Run quietly (suppress normal stderr output).\n",
            "         -s:  Write all potential genes (with scores) to the selected file.\n",
            "         -t:  Write a training file (if none exists); otherwise, read and use\n",
            "              the specified training file.\n",
            "         -v:  Print version number and exit.\n",
            "\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              ""
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 4
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "fR3WMO-puFBc",
        "outputId": "21594b02-e1f2-46a6-849b-722d6afdd2a5"
      },
      "source": [
        "#RFPlasmid requires the randomForest package. Probably already installed on most distributions but surprisingly not on Google Colab\n",
        "%%shell\n",
        "Rscript -e 'install.packages(\"randomForest\")'"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Installing package into ‘/usr/local/lib/R/site-library’\n",
            "(as ‘lib’ is unspecified)\n",
            "trying URL 'https://cran.rstudio.com/src/contrib/randomForest_4.6-14.tar.gz'\n",
            "Content type 'application/x-gzip' length 80074 bytes (78 KB)\n",
            "==================================================\n",
            "downloaded 78 KB\n",
            "\n",
            "* installing *source* package ‘randomForest’ ...\n",
            "** package ‘randomForest’ successfully unpacked and MD5 sums checked\n",
            "** using staged installation\n",
            "** libs\n",
            "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-vB4ZXq/r-base-4.1.0=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c classTree.c -o classTree.o\n",
            "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-vB4ZXq/r-base-4.1.0=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c init.c -o init.o\n",
            "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-vB4ZXq/r-base-4.1.0=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c regTree.c -o regTree.o\n",
            "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-vB4ZXq/r-base-4.1.0=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c regrf.c -o regrf.o\n",
            "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-vB4ZXq/r-base-4.1.0=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c rf.c -o rf.o\n",
            "gfortran -fno-optimize-sibling-calls  -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-vB4ZXq/r-base-4.1.0=. -fstack-protector-strong  -c rfsub.f -o rfsub.o\n",
            "gcc -std=gnu99 -I\"/usr/share/R/include\" -DNDEBUG      -fpic  -g -O2 -fdebug-prefix-map=/build/r-base-vB4ZXq/r-base-4.1.0=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -g  -c rfutils.c -o rfutils.o\n",
            "gcc -std=gnu99 -shared -L/usr/lib/R/lib -Wl,-Bsymbolic-functions -Wl,-z,relro -o randomForest.so classTree.o init.o regTree.o regrf.o rf.o rfsub.o rfutils.o -lgfortran -lm -lquadmath -L/usr/lib/R/lib -lR\n",
            "installing to /usr/local/lib/R/site-library/00LOCK-randomForest/00new/randomForest/libs\n",
            "** R\n",
            "** data\n",
            "** inst\n",
            "** byte-compile and prepare package for lazy loading\n",
            "** help\n",
            "*** installing help indices\n",
            "** building package indices\n",
            "** testing if installed package can be loaded from temporary location\n",
            "** checking absolute paths in shared objects and dynamic libraries\n",
            "** testing if installed package can be loaded from final location\n",
            "** testing if installed package keeps a record of temporary installation path\n",
            "* DONE (randomForest)\n",
            "\n",
            "The downloaded source packages are in\n",
            "\t‘/tmp/RtmpW4lJ53/downloaded_packages’\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              ""
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 5
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "-Ksuhfp6eSib",
        "outputId": "00e0dc1d-4cab-44ea-b5e9-438616b76008"
      },
      "source": [
        " #installing RFPlasmid using pip\n",
        " %%shell\n",
        " pip3 install rfplasmid"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Collecting rfplasmid\n",
            "\u001b[?25l  Downloading https://files.pythonhosted.org/packages/79/a9/0a1262a2fe72736c5d43fb75f34d3cf4b25343fe87acc8bc9636600e092d/rfplasmid-0.0.17-py3-none-any.whl (52.1MB)\n",
            "\u001b[K     |████████████████████████████████| 52.1MB 83kB/s \n",
            "\u001b[?25hRequirement already satisfied: matplotlib in /usr/local/lib/python3.7/dist-packages (from rfplasmid) (3.2.2)\n",
            "Collecting pysam\n",
            "\u001b[?25l  Downloading https://files.pythonhosted.org/packages/20/85/335857b9888f6d9a13b03a8f21b0a6228b180c361631d9d70e7be3e22163/pysam-0.16.0.1-cp37-cp37m-manylinux1_x86_64.whl (9.9MB)\n",
            "\u001b[K     |████████████████████████████████| 9.9MB 23.2MB/s \n",
            "\u001b[?25hRequirement already satisfied: pandas in /usr/local/lib/python3.7/dist-packages (from rfplasmid) (1.1.5)\n",
            "Collecting checkm-genome\n",
            "\u001b[?25l  Downloading https://files.pythonhosted.org/packages/42/91/ad3bc30ca4fd15b831b7e637d59c7fbae6279077e5a4f563dca2713be2db/checkm-genome-1.1.3.tar.gz (118kB)\n",
            "\u001b[K     |████████████████████████████████| 122kB 53.4MB/s \n",
            "\u001b[?25hCollecting biopython\n",
            "\u001b[?25l  Downloading https://files.pythonhosted.org/packages/5a/42/de1ed545df624180b84c613e5e4de4848f72989ce5846a74af6baa0737b9/biopython-1.79-cp37-cp37m-manylinux_2_5_x86_64.manylinux1_x86_64.whl (2.3MB)\n",
            "\u001b[K     |████████████████████████████████| 2.3MB 34.1MB/s \n",
            "\u001b[?25hRequirement already satisfied: numpy in /usr/local/lib/python3.7/dist-packages (from rfplasmid) (1.19.5)\n",
            "Requirement already satisfied: pyparsing!=2.0.4,!=2.1.2,!=2.1.6,>=2.0.1 in /usr/local/lib/python3.7/dist-packages (from matplotlib->rfplasmid) (2.4.7)\n",
            "Requirement already satisfied: python-dateutil>=2.1 in /usr/local/lib/python3.7/dist-packages (from matplotlib->rfplasmid) (2.8.1)\n",
            "Requirement already satisfied: cycler>=0.10 in /usr/local/lib/python3.7/dist-packages (from matplotlib->rfplasmid) (0.10.0)\n",
            "Requirement already satisfied: kiwisolver>=1.0.1 in /usr/local/lib/python3.7/dist-packages (from matplotlib->rfplasmid) (1.3.1)\n",
            "Requirement already satisfied: pytz>=2017.2 in /usr/local/lib/python3.7/dist-packages (from pandas->rfplasmid) (2018.9)\n",
            "Requirement already satisfied: scipy>=0.19.1 in /usr/local/lib/python3.7/dist-packages (from checkm-genome->rfplasmid) (1.4.1)\n",
            "Collecting dendropy>=4.4.0\n",
            "\u001b[?25l  Downloading https://files.pythonhosted.org/packages/f9/10/125c181b1d97ffc4661a60ec897cfe058dc46cb53900d807819464c3510f/DendroPy-4.5.2.tar.gz (15.2MB)\n",
            "\u001b[K     |████████████████████████████████| 15.2MB 271kB/s \n",
            "\u001b[?25hRequirement already satisfied: setuptools in /usr/local/lib/python3.7/dist-packages (from checkm-genome->rfplasmid) (57.0.0)\n",
            "Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.7/dist-packages (from python-dateutil>=2.1->matplotlib->rfplasmid) (1.15.0)\n",
            "Building wheels for collected packages: checkm-genome, dendropy\n",
            "  Building wheel for checkm-genome (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for checkm-genome: filename=checkm_genome-1.1.3-cp37-none-any.whl size=159041 sha256=47a332f8d20ece725a3fbfa8a6e186e3559c6acebb06174a429cf083f8ca3ea1\n",
            "  Stored in directory: /root/.cache/pip/wheels/e9/e6/23/9799e23dc1859a5211d818599458284ff739f1e29d9dfbff4d\n",
            "  Building wheel for dendropy (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for dendropy: filename=DendroPy-4.5.2-cp37-none-any.whl size=453171 sha256=5b48115e7fbaf486d3ec6aa54d9e42c5d4c98fae3d310483617eb98a824d869d\n",
            "  Stored in directory: /root/.cache/pip/wheels/56/4c/3e/b70f6ee1aebabe4206fd2cc4f3d08e21c31a50f98d8dde0701\n",
            "Successfully built checkm-genome dendropy\n",
            "Installing collected packages: pysam, dendropy, checkm-genome, biopython, rfplasmid\n",
            "Successfully installed biopython-1.79 checkm-genome-1.1.3 dendropy-4.5.2 pysam-0.16.0.1 rfplasmid-0.0.17\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              ""
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 6
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "C-5NU2rDhMrz",
        "outputId": "02ffaef0-fb74-462b-e76a-1e21e7b546f1"
      },
      "source": [
        "#getting checkM databases. This is important and often overlooked\n",
        "%%shell\n",
        "mkdir /usr/local/checkm_data\n",
        "cd /usr/local/checkm_data/\n",
        "wget https://data.ace.uq.edu.au/public/CheckM_databases/checkm_data_2015_01_16.tar.gz\n",
        "tar xzvf checkm_data_2015_01_16.tar.gz\n",
        "checkm data setRoot /usr/local/checkm_data/"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "--2021-06-07 13:59:57--  https://data.ace.uq.edu.au/public/CheckM_databases/checkm_data_2015_01_16.tar.gz\n",
            "Resolving data.ace.uq.edu.au (data.ace.uq.edu.au)... 130.102.99.4\n",
            "Connecting to data.ace.uq.edu.au (data.ace.uq.edu.au)|130.102.99.4|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 288590617 (275M) [application/x-gzip]\n",
            "Saving to: ‘checkm_data_2015_01_16.tar.gz’\n",
            "\n",
            "checkm_data_2015_01 100%[===================>] 275.22M  9.44MB/s    in 52s     \n",
            "\n",
            "2021-06-07 14:00:51 (5.30 MB/s) - ‘checkm_data_2015_01_16.tar.gz’ saved [288590617/288590617]\n",
            "\n",
            "./\n",
            "./distributions/\n",
            "./distributions/cd_dist.txt\n",
            "./distributions/gc_dist.txt\n",
            "./distributions/td_dist.txt\n",
            "./genome_tree/\n",
            "./genome_tree/genome_tree.derep.txt\n",
            "./genome_tree/genome_tree.metadata.tsv\n",
            "./genome_tree/missing_duplicate_genes_97.tsv\n",
            "./genome_tree/missing_duplicate_genes_50.tsv\n",
            "./genome_tree/genome_tree_reduced.refpkg/\n",
            "./genome_tree/genome_tree_reduced.refpkg/CONTENTS.json\n",
            "./genome_tree/genome_tree_reduced.refpkg/genome_tree.fasta\n",
            "./genome_tree/genome_tree_reduced.refpkg/genome_tree.log\n",
            "./genome_tree/genome_tree_reduced.refpkg/genome_tree.tre\n",
            "./genome_tree/genome_tree_reduced.refpkg/phylo_modelJqWx6_.json\n",
            "./genome_tree/genome_tree_full.refpkg/\n",
            "./genome_tree/genome_tree_full.refpkg/CONTENTS.json\n",
            "./genome_tree/genome_tree_full.refpkg/phylo_modelEcOyPk.json\n",
            "./genome_tree/genome_tree_full.refpkg/genome_tree.fasta\n",
            "./genome_tree/genome_tree_full.refpkg/genome_tree.log\n",
            "./genome_tree/genome_tree_full.refpkg/genome_tree.tre\n",
            "./genome_tree/genome_tree.taxonomy.tsv\n",
            "./hmms/\n",
            "./hmms/phylo.hmm\n",
            "./hmms/phylo.hmm.ssi\n",
            "./hmms/checkm.hmm\n",
            "./hmms/checkm.hmm.ssi\n",
            "./hmms_ssu/\n",
            "./hmms_ssu/SSU_archaea.hmm\n",
            "./hmms_ssu/SSU_bacteria.hmm\n",
            "./hmms_ssu/SSU_euk.hmm\n",
            "./hmms_ssu/createHMMs.py\n",
            "./img/\n",
            "./img/img_metadata.tsv\n",
            "./pfam/\n",
            "./pfam/Pfam-A.hmm.dat\n",
            "./pfam/tigrfam2pfam.tsv\n",
            "./selected_marker_sets.tsv\n",
            "./taxon_marker_sets.tsv\n",
            "./test_data/\n",
            "./test_data/637000110.fna\n",
            "./.dmanifest\n",
            "It seems that the CheckM data folder has not been set yet or has been removed. Please run 'checkm data setRoot'.\n",
            "\n",
            "Path [/root/.checkm] does not exist so I will attempt to create it\n",
            "Path [/root/.checkm] has been created and you have permission to write to this folder.\n",
            "(re) creating manifest file (please be patient).\n",
            "[2021-06-07 14:01:05] INFO: CheckM v1.1.3\n",
            "[2021-06-07 14:01:05] INFO: checkm data setRoot /usr/local/checkm_data/\n",
            "[2021-06-07 14:01:05] INFO: [CheckM - data] Check for database updates. [setRoot]\n",
            "\n",
            "Path [/usr/local/checkm_data] exists and you have permission to write to this folder.\n",
            "(re) creating manifest file (please be patient).\n",
            "\n",
            "Path [/usr/local/checkm_data] exists and you have permission to write to this folder.\n",
            "(re) creating manifest file (please be patient).\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              ""
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 7
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "id": "NJc933YaelVt",
        "outputId": "0cf44f0f-5e8f-480d-a29c-8aa485e77e75"
      },
      "source": [
        "# Initialize RFPlasmid databases\n",
        "%%shell\n",
        "rfplasmid --initialize #use --initialize first to check if RFPlasmid databases are present and formatted, if not, download them\n",
        "# This may end with an error because of Google Colab pecularities and not being and interactive shell. It still works. "
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Warning: Jellyfish is not installed. Not recommended\n",
            "Downloading plasmid databases.. Be patient. You may have to run this script with sudo\n",
            "This script needs wget and diamond\n",
            "Downloading databases from http://klif.uu.nl/download/plasmid_db/\n",
            "Please be patient. This is a large download\n",
            "\n",
            "Download plasmid genes databases\n",
            "--2021-06-07 14:01:24--  http://klif.uu.nl/download/plasmid_db/plasmiddb_cge.faa\n",
            "Resolving klif.uu.nl (klif.uu.nl)... 131.211.18.5\n",
            "Connecting to klif.uu.nl (klif.uu.nl)|131.211.18.5|:80... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 299612 (293K)\n",
            "Saving to: ‘plasmiddb_cge.faa’\n",
            "\n",
            "plasmiddb_cge.faa   100%[===================>] 292.59K   869KB/s    in 0.3s    \n",
            "\n",
            "2021-06-07 14:01:25 (869 KB/s) - ‘plasmiddb_cge.faa’ saved [299612/299612]\n",
            "\n",
            "--2021-06-07 14:01:25--  http://klif.uu.nl/download/plasmid_db/plasmiddb_total.faa\n",
            "Resolving klif.uu.nl (klif.uu.nl)... 131.211.18.5\n",
            "Connecting to klif.uu.nl (klif.uu.nl)|131.211.18.5|:80... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 645258917 (615M)\n",
            "Saving to: ‘plasmiddb_total.faa’\n",
            "\n",
            "plasmiddb_total.faa 100%[===================>] 615.37M  16.5MB/s    in 34s     \n",
            "\n",
            "2021-06-07 14:01:59 (18.1 MB/s) - ‘plasmiddb_total.faa’ saved [645258917/645258917]\n",
            "\n",
            "Indexing files\n",
            "diamond v0.9.24.125 | by Benjamin Buchfink <buchfink@gmail.com>\n",
            "Licensed under the GNU GPL <https://www.gnu.org/licenses/gpl.txt>\n",
            "Check http://github.com/bbuchfink/diamond for updates.\n",
            "\n",
            "#CPU threads: 2\n",
            "Scoring parameters: (Matrix=BLOSUM62 Lambda=0.267 K=0.041 Penalties=11/1)\n",
            "Database file: plasmiddb_cge.faa\n",
            "Opening the database file...  [0.000228s]\n",
            "Loading sequences...  [0.001699s]\n",
            "Masking sequences...  [0.055726s]\n",
            "Writing sequences...  [0.000509s]\n",
            "Hashing sequences...  [0.000134s]\n",
            "Loading sequences...  [1.9e-05s]\n",
            "Writing trailer...  [3.1e-05s]\n",
            "Closing the input file...  [2.1e-05s]\n",
            "Closing the database file...  [2.9e-05s]\n",
            "Database hash = ca4f143292c309e1dc6002fd9199893f\n",
            "Processed 1140 sequences, 260876 letters.\n",
            "Total time = 0.058632s\n",
            "diamond v0.9.24.125 | by Benjamin Buchfink <buchfink@gmail.com>\n",
            "Licensed under the GNU GPL <https://www.gnu.org/licenses/gpl.txt>\n",
            "Check http://github.com/bbuchfink/diamond for updates.\n",
            "\n",
            "#CPU threads: 2\n",
            "Scoring parameters: (Matrix=BLOSUM62 Lambda=0.267 K=0.041 Penalties=11/1)\n",
            "Database file: plasmiddb_total.faa\n",
            "Opening the database file...  [0.00057s]\n",
            "Loading sequences...  [4.07298s]\n",
            "Masking sequences...  [107.495s]\n",
            "Writing sequences...  [2.09658s]\n",
            "Hashing sequences...  [0.217461s]\n",
            "Loading sequences...  [3.1e-05s]\n",
            "Writing trailer...  [0.062587s]\n",
            "Closing the input file...  [4e-05s]\n",
            "Closing the database file...  [6.7e-05s]\n",
            "Database hash = 948611271e55ddde24f77998a94ef865\n",
            "Processed 1832663 sequences, 516585005 letters.\n",
            "Total time = 113.946s\n",
            "Done\n",
            "\n",
            " installed. Please try rfplasmid again\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "error",
          "ename": "CalledProcessError",
          "evalue": "ignored",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mCalledProcessError\u001b[0m                        Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-8-675804dce71c>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mget_ipython\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrun_cell_magic\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'shell'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m''\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'rfplasmid --initialize #use --initialize first to check if RFPlasmid databases are present and formatted, if not, download them\\n# This may end with an error because of Google Colab pecularities and not being and interactive shell. It still works. '\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
            "\u001b[0;32m/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py\u001b[0m in \u001b[0;36mrun_cell_magic\u001b[0;34m(self, magic_name, line, cell)\u001b[0m\n\u001b[1;32m   2115\u001b[0m             \u001b[0mmagic_arg_s\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvar_expand\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mline\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstack_depth\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2116\u001b[0m             \u001b[0;32mwith\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbuiltin_trap\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2117\u001b[0;31m                 \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmagic_arg_s\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcell\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2118\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2119\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.7/dist-packages/google/colab/_system_commands.py\u001b[0m in \u001b[0;36m_shell_cell_magic\u001b[0;34m(args, cmd)\u001b[0m\n\u001b[1;32m    111\u001b[0m   \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_run_command\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcmd\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mclear_streamed_output\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    112\u001b[0m   \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mparsed_args\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mignore_errors\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 113\u001b[0;31m     \u001b[0mresult\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcheck_returncode\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    114\u001b[0m   \u001b[0;32mreturn\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    115\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;32m/usr/local/lib/python3.7/dist-packages/google/colab/_system_commands.py\u001b[0m in \u001b[0;36mcheck_returncode\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    137\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreturncode\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    138\u001b[0m       raise subprocess.CalledProcessError(\n\u001b[0;32m--> 139\u001b[0;31m           returncode=self.returncode, cmd=self.args, output=self.output)\n\u001b[0m\u001b[1;32m    140\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    141\u001b[0m   \u001b[0;32mdef\u001b[0m \u001b[0m_repr_pretty_\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mp\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcycle\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m  \u001b[0;31m# pylint:disable=unused-argument\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mCalledProcessError\u001b[0m: Command 'rfplasmid --initialize #use --initialize first to check if RFPlasmid databases are present and formatted, if not, download them\n# This may end with an error because of Google Colab pecularities and not being and interactive shell. It still works. ' returned non-zero exit status 1."
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "_H9UEYa_jyXt",
        "outputId": "ff8ec0bc-fbca-48aa-89ea-5ea137acea35"
      },
      "source": [
        "#get an example folder which is part of the rfplasmid github\n",
        "%%shell\n",
        "git clone https://github.com/aldertzomer/RFPlasmid.git"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Cloning into 'RFPlasmid'...\n",
            "remote: Enumerating objects: 305, done.\u001b[K\n",
            "remote: Counting objects: 100% (148/148), done.\u001b[K\n",
            "remote: Compressing objects: 100% (148/148), done.\u001b[K\n",
            "remote: Total 305 (delta 91), reused 0 (delta 0), pack-reused 157\u001b[K\n",
            "Receiving objects: 100% (305/305), 51.40 MiB | 34.18 MiB/s, done.\n",
            "Resolving deltas: 100% (159/159), done.\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              ""
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 9
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "0qOqhVWUe0Yl",
        "outputId": "7e501324-6fdb-43d3-cd08-172939aee456"
      },
      "source": [
        "# Does RFPlasmid work? \n",
        "%%shell\n",
        "# Here we run rfplasmid on the examplefolder from github. Get your own folder from google drive using the example below\n",
        "# This is a run without Jellyfish. For a bit of extra speedup, Jellyfish can be installed, however for the few genomes in the example it won't matter that much.\n",
        "rfplasmid --species Campylobacter --input RFPlasmid/example --threads 8 --out outputfolder6 --debug"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Warning: Jellyfish is not installed. Not recommended\n",
            "\n",
            "\n",
            "Start RFPlasmid, version 0.0.17\n",
            "copy original contig names\n",
            "cleanup contig names\n",
            "start Checkm\n",
            "    Finished processing 4 of 4 (100.00%) bins.\n",
            "    Finished parsing hits for 4 of 4 (100.00%) bins.\n",
            "    Finished processing 4 of 4 (100.00%) bins.\n",
            "    Finished processing 4 of 4 (100.00%) bins.\n",
            "    Finished parsing hits for 4 of 4 (100.00%) bins.\n",
            "Checkm done\n",
            "start blast plasmiddb\n",
            "/usr/local/lib/python3.7/dist-packages/RFPlasmid/rfplasmid.py:154: FutureWarning: Columnar iteration over characters will be deprecated in future releases.\n",
            "  df1['contig'], df1['gene'] = df1['contig_gene'].str.split('_', 1).str\n",
            "/usr/local/lib/python3.7/dist-packages/RFPlasmid/rfplasmid.py:154: FutureWarning: Columnar iteration over characters will be deprecated in future releases.\n",
            "  df1['contig'], df1['gene'] = df1['contig_gene'].str.split('_', 1).str\n",
            "/usr/local/lib/python3.7/dist-packages/RFPlasmid/rfplasmid.py:154: FutureWarning: Columnar iteration over characters will be deprecated in future releases.\n",
            "  df1['contig'], df1['gene'] = df1['contig_gene'].str.split('_', 1).str\n",
            "/usr/local/lib/python3.7/dist-packages/RFPlasmid/rfplasmid.py:154: FutureWarning: Columnar iteration over characters will be deprecated in future releases.\n",
            "  df1['contig'], df1['gene'] = df1['contig_gene'].str.split('_', 1).str\n",
            "blast plasmid genes done\n",
            "start blast cge\n",
            "blast cge genes done\n",
            "count genes per contig done\n",
            "/usr/local/lib/python3.7/dist-packages/RFPlasmid/rfplasmid.py:213: FutureWarning: Columnar iteration over characters will be deprecated in future releases.\n",
            "  df['contig'], df['gene'] = df['contig_gene'].str.split('_', 1).str\n",
            "count SCMs per contig done\n",
            "Start kmer count\n",
            "Standard kmer-counting method\n",
            "Kmer count standard method ready\n",
            "Count contig length\n",
            "count contig length ; done\n",
            "Start merging contig files\n",
            "Create df_files ; done\n",
            "Start to add blast output to dataframe\n",
            "CSV file from dataframe ready in file outputdataframe.csv\n",
            "debug mode, no cleaning up\n",
            "Start R\n",
            "RFPlasmid prediction mode started\n",
            "\n",
            "R version 4.1.0 (2021-05-18) -- \"Camp Pontanezen\"\n",
            "Copyright (C) 2021 The R Foundation for Statistical Computing\n",
            "Platform: x86_64-pc-linux-gnu (64-bit)\n",
            "\n",
            "R is free software and comes with ABSOLUTELY NO WARRANTY.\n",
            "You are welcome to redistribute it under certain conditions.\n",
            "Type 'license()' or 'licence()' for distribution details.\n",
            "\n",
            "  Natural language support but running in an English locale\n",
            "\n",
            "R is a collaborative project with many contributors.\n",
            "Type 'contributors()' for more information and\n",
            "'citation()' on how to cite R or R packages in publications.\n",
            "\n",
            "Type 'demo()' for some demos, 'help()' for on-line help, or\n",
            "'help.start()' for an HTML browser interface to help.\n",
            "Type 'q()' to quit R.\n",
            "\n",
            "> #classifying\n",
            "> Args <- commandArgs(TRUE)\n",
            "> library(randomForest)\n",
            "randomForest 4.6-14\n",
            "Type rfNews() to see new features/changes/bug fixes.\n",
            "> model<-paste(Args[2],\"/\",Args[1],\".rfo\",sep=\"\")\n",
            "> load(model)\n",
            "> df<-read.table(\"outputdataframe.csv\", sep=\",\", header=T, row.names=1)\n",
            "> featurematrix <-df\n",
            "> featurematrix$contigID <- NULL\n",
            "> featurematrix$genome <- NULL\n",
            "> #notice the missing featurevector generation and the use of predict\n",
            "> rf.classifications <- predict(rf,featurematrix)\n",
            "> rf.votes <- predict(rf,featurematrix, type=\"vote\")\n",
            "> combined1 <- merge(rf.classifications,rf.votes,by=\"row.names\")\n",
            "> row.names(combined1)<-combined1$Row.names\n",
            "> combined1$Row.names <-NULL\n",
            "> colnames(combined1) <- c(\"prediction\", \"votes chromosomal\", \"votes plasmid\")\n",
            "> output <- merge(combined1,df,by=\"row.names\")\n",
            "> row.names(output)<-output$Row.names\n",
            "> output$Row.names <-NULL\n",
            "> write.csv(output[1:4], file=\"prediction.csv\")\n",
            "> write.csv(output, file=\"prediction_full.csv\")\n",
            "> save.image(\"classification.RData\")\n",
            "> \n",
            "Prediction done\n",
            "RFPlasmid output can be found in directory outputfolder6\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              ""
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 10
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "v_SnIfXqw3uM",
        "outputId": "6e2e9072-d66e-49e3-d1f4-513159ca51b7"
      },
      "source": [
        "# It works!\n",
        "%%shell\n",
        "cat outputfolder6/prediction.csv"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "\"\",\"prediction\",\"votes chromosomal\",\"votes plasmid\",\"contigID\"\n",
            "\"cGCA_000009085.1.assembled_1\",\"c\",0.9986,0.0014,\"NODE_1_length_481500_cov_37.2278\"\n",
            "\"cGCA_000009085.1.assembled_10\",\"c\",0.9358,0.0642,\"NODE_10_length_42628_cov_38.7994\"\n",
            "\"cGCA_000009085.1.assembled_11\",\"c\",0.9762,0.0238,\"NODE_11_length_41206_cov_37.1006\"\n",
            "\"cGCA_000009085.1.assembled_12\",\"c\",0.9848,0.0152,\"NODE_12_length_38888_cov_35.6564\"\n",
            "\"cGCA_000009085.1.assembled_13\",\"c\",0.6758,0.3242,\"NODE_13_length_20567_cov_36.6903\"\n",
            "\"cGCA_000009085.1.assembled_14\",\"c\",0.817,0.183,\"NODE_14_length_15123_cov_38.583\"\n",
            "\"cGCA_000009085.1.assembled_15\",\"c\",0.7266,0.2734,\"NODE_15_length_13284_cov_38.3662\"\n",
            "\"cGCA_000009085.1.assembled_16\",\"c\",0.955,0.045,\"NODE_16_length_6071_cov_115.105\"\n",
            "\"cGCA_000009085.1.assembled_17\",\"c\",0.7292,0.2708,\"NODE_17_length_4640_cov_35.8843\"\n",
            "\"cGCA_000009085.1.assembled_18\",\"c\",0.8788,0.1212,\"NODE_18_length_2404_cov_37.7646\"\n",
            "\"cGCA_000009085.1.assembled_19\",\"c\",0.864,0.136,\"NODE_19_length_2340_cov_76.0642\"\n",
            "\"cGCA_000009085.1.assembled_2\",\"c\",0.9946,0.0054,\"NODE_2_length_189488_cov_37.5042\"\n",
            "\"cGCA_000009085.1.assembled_20\",\"c\",0.7126,0.2874,\"NODE_20_length_1636_cov_72.6528\"\n",
            "\"cGCA_000009085.1.assembled_21\",\"c\",0.8024,0.1976,\"NODE_21_length_1561_cov_71.6541\"\n",
            "\"cGCA_000009085.1.assembled_22\",\"c\",0.912,0.088,\"NODE_22_length_1462_cov_37.609\"\n",
            "\"cGCA_000009085.1.assembled_23\",\"c\",0.7604,0.2396,\"NODE_23_length_1379_cov_37.7804\"\n",
            "\"cGCA_000009085.1.assembled_24\",\"c\",0.8802,0.1198,\"NODE_24_length_1342_cov_38.9523\"\n",
            "\"cGCA_000009085.1.assembled_25\",\"c\",0.898,0.102,\"NODE_25_length_1041_cov_116.69\"\n",
            "\"cGCA_000009085.1.assembled_26\",\"c\",0.691,0.309,\"NODE_26_length_947_cov_68.8073\"\n",
            "\"cGCA_000009085.1.assembled_27\",\"c\",0.8042,0.1958,\"NODE_27_length_807_cov_82.425\"\n",
            "\"cGCA_000009085.1.assembled_28\",\"c\",0.8204,0.1796,\"NODE_28_length_656_cov_78.1304\"\n",
            "\"cGCA_000009085.1.assembled_29\",\"c\",0.7486,0.2514,\"NODE_29_length_496_cov_70.3631\"\n",
            "\"cGCA_000009085.1.assembled_3\",\"c\",0.9918,0.0082,\"NODE_3_length_154057_cov_37.2165\"\n",
            "\"cGCA_000009085.1.assembled_30\",\"c\",0.717,0.283,\"NODE_30_length_447_cov_33.9937\"\n",
            "\"cGCA_000009085.1.assembled_4\",\"c\",0.9788,0.0212,\"NODE_4_length_115825_cov_37.5027\"\n",
            "\"cGCA_000009085.1.assembled_5\",\"c\",0.9754,0.0246,\"NODE_5_length_104903_cov_36.7652\"\n",
            "\"cGCA_000009085.1.assembled_6\",\"c\",0.9776,0.0224,\"NODE_6_length_104312_cov_37.8467\"\n",
            "\"cGCA_000009085.1.assembled_7\",\"c\",0.93,0.07,\"NODE_7_length_103032_cov_36.98\"\n",
            "\"cGCA_000009085.1.assembled_8\",\"c\",0.9848,0.0152,\"NODE_8_length_90401_cov_37.1984\"\n",
            "\"cGCA_000009085.1.assembled_9\",\"c\",0.978,0.022,\"NODE_9_length_80900_cov_37.7382\"\n",
            "\"cGCA_000015085.1.assembled_1\",\"c\",0.8744,0.1256,\"NODE_1_length_616191_cov_37.3872\"\n",
            "\"cGCA_000015085.1.assembled_10\",\"c\",0.5968,0.4032,\"NODE_10_length_20564_cov_38.5059\"\n",
            "\"cGCA_000015085.1.assembled_11\",\"c\",0.8372,0.1628,\"NODE_11_length_16099_cov_39.5755\"\n",
            "\"cGCA_000015085.1.assembled_12\",\"c\",0.8788,0.1212,\"NODE_12_length_14603_cov_43.757\"\n",
            "\"cGCA_000015085.1.assembled_13\",\"c\",0.819,0.181,\"NODE_13_length_6420_cov_42.2452\"\n",
            "\"cGCA_000015085.1.assembled_14\",\"c\",0.8628,0.1372,\"NODE_14_length_6317_cov_111.849\"\n",
            "\"cGCA_000015085.1.assembled_15\",\"c\",0.8284,0.1716,\"NODE_15_length_5256_cov_56.9302\"\n",
            "\"cGCA_000015085.1.assembled_16\",\"c\",0.8466,0.1534,\"NODE_16_length_4562_cov_45.4322\"\n",
            "\"cGCA_000015085.1.assembled_17\",\"c\",0.755,0.245,\"NODE_17_length_4304_cov_39.8066\"\n",
            "\"cGCA_000015085.1.assembled_18\",\"c\",0.8166,0.1834,\"NODE_18_length_3996_cov_48.6242\"\n",
            "\"cGCA_000015085.1.assembled_19\",\"c\",0.8238,0.1762,\"NODE_19_length_3426_cov_46.1852\"\n",
            "\"cGCA_000015085.1.assembled_2\",\"c\",0.9264,0.0736,\"NODE_2_length_260927_cov_37.5341\"\n",
            "\"cGCA_000015085.1.assembled_20\",\"c\",0.7666,0.2334,\"NODE_20_length_2079_cov_40.9677\"\n",
            "\"cGCA_000015085.1.assembled_21\",\"c\",0.6106,0.3894,\"NODE_21_length_1910_cov_37.3023\"\n",
            "\"cGCA_000015085.1.assembled_22\",\"c\",0.7204,0.2796,\"NODE_22_length_1875_cov_47.1087\"\n",
            "\"cGCA_000015085.1.assembled_23\",\"c\",0.725,0.275,\"NODE_23_length_1535_cov_54.5426\"\n",
            "\"cGCA_000015085.1.assembled_24\",\"c\",0.8142,0.1858,\"NODE_24_length_1135_cov_35.9821\"\n",
            "\"cGCA_000015085.1.assembled_25\",\"c\",0.7238,0.2762,\"NODE_25_length_1042_cov_75.6546\"\n",
            "\"cGCA_000015085.1.assembled_26\",\"c\",0.7366,0.2634,\"NODE_26_length_982_cov_113.319\"\n",
            "\"cGCA_000015085.1.assembled_27\",\"c\",0.6836,0.3164,\"NODE_27_length_815_cov_35.1047\"\n",
            "\"cGCA_000015085.1.assembled_28\",\"c\",0.79,0.21,\"NODE_28_length_758_cov_85.336\"\n",
            "\"cGCA_000015085.1.assembled_29\",\"c\",0.7798,0.2202,\"NODE_29_length_652_cov_103.248\"\n",
            "\"cGCA_000015085.1.assembled_3\",\"c\",0.9118,0.0882,\"NODE_3_length_159966_cov_36.952\"\n",
            "\"cGCA_000015085.1.assembled_30\",\"c\",0.796,0.204,\"NODE_30_length_641_cov_70.4767\"\n",
            "\"cGCA_000015085.1.assembled_31\",\"c\",0.7822,0.2178,\"NODE_31_length_585_cov_332.05\"\n",
            "\"cGCA_000015085.1.assembled_32\",\"c\",0.5826,0.4174,\"NODE_32_length_525_cov_84.3266\"\n",
            "\"cGCA_000015085.1.assembled_33\",\"c\",0.7302,0.2698,\"NODE_33_length_523_cov_76.2955\"\n",
            "\"cGCA_000015085.1.assembled_34\",\"c\",0.7332,0.2668,\"NODE_34_length_503_cov_62.8271\"\n",
            "\"cGCA_000015085.1.assembled_35\",\"c\",0.5958,0.4042,\"NODE_35_length_441_cov_39.419\"\n",
            "\"cGCA_000015085.1.assembled_4\",\"c\",0.8936,0.1064,\"NODE_4_length_157627_cov_37.4275\"\n",
            "\"cGCA_000015085.1.assembled_5\",\"c\",0.8728,0.1272,\"NODE_5_length_155390_cov_37.0545\"\n",
            "\"cGCA_000015085.1.assembled_6\",\"c\",0.892,0.108,\"NODE_6_length_133308_cov_37.4763\"\n",
            "\"cGCA_000015085.1.assembled_7\",\"c\",0.8728,0.1272,\"NODE_7_length_88761_cov_37.182\"\n",
            "\"cGCA_000015085.1.assembled_8\",\"c\",0.8532,0.1468,\"NODE_8_length_46732_cov_37.1925\"\n",
            "\"cGCA_000015085.1.assembled_9\",\"c\",0.839,0.161,\"NODE_9_length_29604_cov_38.1727\"\n",
            "\"pGCA_001717605.1.assembled_1\",\"p\",0.0274,0.9726,\"NODE_1_length_34488_cov_37.4848\"\n",
            "\"pGCA_001717605.1.assembled_10\",\"p\",0.2314,0.7686,\"NODE_10_length_4323_cov_39.9013\"\n",
            "\"pGCA_001717605.1.assembled_11\",\"p\",0.2322,0.7678,\"NODE_11_length_4040_cov_36.2098\"\n",
            "\"pGCA_001717605.1.assembled_12\",\"p\",0.2732,0.7268,\"NODE_12_length_3133_cov_36.6437\"\n",
            "\"pGCA_001717605.1.assembled_13\",\"p\",0.1608,0.8392,\"NODE_13_length_3130_cov_37.2627\"\n",
            "\"pGCA_001717605.1.assembled_14\",\"p\",0.1254,0.8746,\"NODE_14_length_2656_cov_39.3693\"\n",
            "\"pGCA_001717605.1.assembled_15\",\"p\",0.2636,0.7364,\"NODE_15_length_2203_cov_39.5366\"\n",
            "\"pGCA_001717605.1.assembled_16\",\"p\",0.2212,0.7788,\"NODE_16_length_2146_cov_42.8494\"\n",
            "\"pGCA_001717605.1.assembled_17\",\"p\",0.1728,0.8272,\"NODE_17_length_2037_cov_37.2497\"\n",
            "\"pGCA_001717605.1.assembled_18\",\"p\",0.2558,0.7442,\"NODE_18_length_1971_cov_60.8964\"\n",
            "\"pGCA_001717605.1.assembled_19\",\"p\",0.221,0.779,\"NODE_19_length_1939_cov_37.8405\"\n",
            "\"pGCA_001717605.1.assembled_2\",\"p\",0.2804,0.7196,\"NODE_2_length_18869_cov_36.0662\"\n",
            "\"pGCA_001717605.1.assembled_20\",\"p\",0.2458,0.7542,\"NODE_20_length_1862_cov_36.4121\"\n",
            "\"pGCA_001717605.1.assembled_21\",\"p\",0.2402,0.7598,\"NODE_21_length_1856_cov_111.736\"\n",
            "\"pGCA_001717605.1.assembled_22\",\"p\",0.2724,0.7276,\"NODE_22_length_1578_cov_73.7726\"\n",
            "\"pGCA_001717605.1.assembled_23\",\"p\",0.232,0.768,\"NODE_23_length_1460_cov_72.9497\"\n",
            "\"pGCA_001717605.1.assembled_24\",\"p\",0.235,0.765,\"NODE_24_length_1397_cov_71.3457\"\n",
            "\"pGCA_001717605.1.assembled_25\",\"p\",0.2694,0.7306,\"NODE_25_length_1360_cov_42.2303\"\n",
            "\"pGCA_001717605.1.assembled_26\",\"p\",0.2974,0.7026,\"NODE_26_length_1143_cov_82.7303\"\n",
            "\"pGCA_001717605.1.assembled_27\",\"p\",0.359,0.641,\"NODE_27_length_1041_cov_48.4453\"\n",
            "\"pGCA_001717605.1.assembled_28\",\"p\",0.3232,0.6768,\"NODE_28_length_964_cov_110.993\"\n",
            "\"pGCA_001717605.1.assembled_29\",\"p\",0.1346,0.8654,\"NODE_29_length_847_cov_35.7653\"\n",
            "\"pGCA_001717605.1.assembled_3\",\"p\",0.1226,0.8774,\"NODE_3_length_17562_cov_38.0518\"\n",
            "\"pGCA_001717605.1.assembled_30\",\"p\",0.3518,0.6482,\"NODE_30_length_788_cov_67.4145\"\n",
            "\"pGCA_001717605.1.assembled_31\",\"p\",0.3468,0.6532,\"NODE_31_length_717_cov_73.7458\"\n",
            "\"pGCA_001717605.1.assembled_32\",\"p\",0.2266,0.7734,\"NODE_32_length_677_cov_87.7509\"\n",
            "\"pGCA_001717605.1.assembled_33\",\"p\",0.149,0.851,\"NODE_33_length_661_cov_73.0337\"\n",
            "\"pGCA_001717605.1.assembled_34\",\"p\",0.3234,0.6766,\"NODE_34_length_531_cov_76.1535\"\n",
            "\"pGCA_001717605.1.assembled_35\",\"p\",0.3426,0.6574,\"NODE_35_length_530_cov_126.702\"\n",
            "\"pGCA_001717605.1.assembled_36\",\"p\",0.384,0.616,\"NODE_36_length_409_cov_40.0674\"\n",
            "\"pGCA_001717605.1.assembled_37\",\"p\",0.433,0.567,\"NODE_37_length_264_cov_33.8175\"\n",
            "\"pGCA_001717605.1.assembled_38\",\"p\",0.4044,0.5956,\"NODE_38_length_261_cov_38.9179\"\n",
            "\"pGCA_001717605.1.assembled_39\",\"p\",0.437,0.563,\"NODE_39_length_205_cov_113.423\"\n",
            "\"pGCA_001717605.1.assembled_4\",\"p\",0.0672,0.9328,\"NODE_4_length_13557_cov_37.1716\"\n",
            "\"pGCA_001717605.1.assembled_5\",\"p\",0.0844,0.9156,\"NODE_5_length_10684_cov_35.7406\"\n",
            "\"pGCA_001717605.1.assembled_6\",\"p\",0.1466,0.8534,\"NODE_6_length_9379_cov_38.8701\"\n",
            "\"pGCA_001717605.1.assembled_7\",\"p\",0.1192,0.8808,\"NODE_7_length_8648_cov_45.3422\"\n",
            "\"pGCA_001717605.1.assembled_8\",\"p\",0.2156,0.7844,\"NODE_8_length_8151_cov_45.6264\"\n",
            "\"pGCA_001717605.1.assembled_9\",\"p\",0.4318,0.5682,\"NODE_9_length_6686_cov_39.1575\"\n",
            "\"pGCA_001936355.1.assembled_1\",\"p\",0.2788,0.7212,\"NODE_1_length_27010_cov_37.3079\"\n",
            "\"pGCA_001936355.1.assembled_10\",\"p\",0.1746,0.8254,\"NODE_10_length_2951_cov_45.2666\"\n",
            "\"pGCA_001936355.1.assembled_11\",\"p\",0.198,0.802,\"NODE_11_length_2878_cov_37.2897\"\n",
            "\"pGCA_001936355.1.assembled_12\",\"p\",0.1274,0.8726,\"NODE_12_length_2656_cov_42.3705\"\n",
            "\"pGCA_001936355.1.assembled_13\",\"p\",0.2652,0.7348,\"NODE_13_length_2328_cov_52.9641\"\n",
            "\"pGCA_001936355.1.assembled_14\",\"p\",0.3528,0.6472,\"NODE_14_length_2167_cov_47.6529\"\n",
            "\"pGCA_001936355.1.assembled_15\",\"p\",0.255,0.745,\"NODE_15_length_1951_cov_39.4984\"\n",
            "\"pGCA_001936355.1.assembled_16\",\"p\",0.2612,0.7388,\"NODE_16_length_1706_cov_81.2951\"\n",
            "\"pGCA_001936355.1.assembled_17\",\"p\",0.281,0.719,\"NODE_17_length_1565_cov_39.436\"\n",
            "\"pGCA_001936355.1.assembled_18\",\"p\",0.2106,0.7894,\"NODE_18_length_1337_cov_66.1504\"\n",
            "\"pGCA_001936355.1.assembled_19\",\"p\",0.3124,0.6876,\"NODE_19_length_1219_cov_42.4826\"\n",
            "\"pGCA_001936355.1.assembled_2\",\"p\",0.1472,0.8528,\"NODE_2_length_15174_cov_37.4556\"\n",
            "\"pGCA_001936355.1.assembled_20\",\"p\",0.3334,0.6666,\"NODE_20_length_1037_cov_43.5495\"\n",
            "\"pGCA_001936355.1.assembled_21\",\"p\",0.3328,0.6672,\"NODE_21_length_959_cov_30.9591\"\n",
            "\"pGCA_001936355.1.assembled_22\",\"p\",0.3214,0.6786,\"NODE_22_length_946_cov_38.8327\"\n",
            "\"pGCA_001936355.1.assembled_23\",\"p\",0.3446,0.6554,\"NODE_23_length_931_cov_40.7226\"\n",
            "\"pGCA_001936355.1.assembled_24\",\"p\",0.307,0.693,\"NODE_24_length_860_cov_28.7394\"\n",
            "\"pGCA_001936355.1.assembled_25\",\"p\",0.137,0.863,\"NODE_25_length_847_cov_34.4931\"\n",
            "\"pGCA_001936355.1.assembled_26\",\"p\",0.288,0.712,\"NODE_26_length_837_cov_79.1141\"\n",
            "\"pGCA_001936355.1.assembled_27\",\"p\",0.394,0.606,\"NODE_27_length_683_cov_114.092\"\n",
            "\"pGCA_001936355.1.assembled_28\",\"p\",0.2268,0.7732,\"NODE_28_length_677_cov_79.4309\"\n",
            "\"pGCA_001936355.1.assembled_29\",\"p\",0.1508,0.8492,\"NODE_29_length_663_cov_81.2687\"\n",
            "\"pGCA_001936355.1.assembled_3\",\"p\",0.0834,0.9166,\"NODE_3_length_10682_cov_34.3131\"\n",
            "\"pGCA_001936355.1.assembled_30\",\"p\",0.3186,0.6814,\"NODE_30_length_645_cov_39.9305\"\n",
            "\"pGCA_001936355.1.assembled_31\",\"p\",0.4538,0.5462,\"NODE_31_length_536_cov_78.8411\"\n",
            "\"pGCA_001936355.1.assembled_32\",\"p\",0.3834,0.6166,\"NODE_32_length_519_cov_42.6556\"\n",
            "\"pGCA_001936355.1.assembled_33\",\"p\",0.428,0.572,\"NODE_33_length_335_cov_67.25\"\n",
            "\"pGCA_001936355.1.assembled_34\",\"p\",0.375,0.625,\"NODE_34_length_298_cov_77.1462\"\n",
            "\"pGCA_001936355.1.assembled_4\",\"p\",0.0996,0.9004,\"NODE_4_length_10020_cov_39.5831\"\n",
            "\"pGCA_001936355.1.assembled_5\",\"p\",0.1114,0.8886,\"NODE_5_length_6161_cov_73.2113\"\n",
            "\"pGCA_001936355.1.assembled_6\",\"p\",0.111,0.889,\"NODE_6_length_5494_cov_35.8522\"\n",
            "\"pGCA_001936355.1.assembled_7\",\"p\",0.2004,0.7996,\"NODE_7_length_3173_cov_41.4064\"\n",
            "\"pGCA_001936355.1.assembled_8\",\"p\",0.3386,0.6614,\"NODE_8_length_2990_cov_37.6706\"\n",
            "\"pGCA_001936355.1.assembled_9\",\"p\",0.2456,0.7544,\"NODE_9_length_2982_cov_53.5765\"\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              ""
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 11
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "8Q9PSHrgwhB_"
      },
      "source": [
        "## How to run it on your own folder on Google Drive. \n",
        "This is experimental and not that well tested. The Google Colab script was only there to showcase the installation procedure. "
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "ngWKq9d0vlJ7"
      },
      "source": [
        "# Make a mountpoint for your google drive\n",
        "%%shell\n",
        "mkdir  gdrive"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "kPHDnCRuv3Nu"
      },
      "source": [
        "# Mount google drive. You will need to authorize access using the link. \n",
        "from google.colab import drive\n",
        "drive.mount('gdrive', force_remount=True)"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "MvOdtCoOwCZP"
      },
      "source": [
        "# Run RFPlasmid on a folder in your Google Drive. The folder should be in the root of your google drive and be called rfplasmid and should contain .fasta files\n",
        "%%shell\n",
        "ls gdrive/rfplasmid\n",
        "rfplasmid --species Campylobacter --input gdrive/rfplasmid --threads 8 --out gdrive/rfplasmidoutput --debug"
      ],
      "execution_count": null,
      "outputs": []
    }
  ]
}